filebeat/prospector/prospector.go

```diff
@@ -195,21 +210,71 @@ func (p *Prospector) updateState(event *input.Event) error {
 	return nil
 }
 
+// Stop stops the prospector and with it all harvesters
+//
+// The shutdown order is as follwoing
+// - stop run and scanning
+// - wait until last scan finishes to make sure no new harvesters are added
+// - stop harvesters
+// - wait until all harvester finished
+// - stop communication channel
+// - wait on internal waitgroup to make sure all prospector go routines are stopped
+// - wait until all events are forwarded to the spooler
 func (p *Prospector) Stop() {
+	// Stop scanning and wait for completion
+	close(p.runDone)
+	p.wg.Wait()
+}
+
+func (p *Prospector) stop() {
+	defer p.wg.Done()
+
 	logp.Info("Stopping Prospector: %v", p.ID())
-	close(p.done)
+
+	// In case of once, it will be waited until harvesters close itself
+	if p.Once {
+		p.registry.waitForCompletion()
+	}
+
+	// Wait for finishing of the running prospectors
+	// This ensure no new harvesters are added.
+	p.runWg.Wait()
+
+	// Stop all harvesters
+	// In case the beatDone channel is closed, this will not wait for completion
+	// Otherwise Stop will wait until output is complete
+	p.registry.Stop()
+
+	// Waits on stopping all harvesters to make sure all events made it into the channel
+	p.waitEvents()
+}
+
+// Wait for completion of sending events
+func (p *Prospector) waitEvents() {
+
+	done := make(chan struct{})
+	go func() {
+		p.eventCounter.Wait()
+		close(done)
+	}()
+
+	select {
+	case <-done:
+		close(p.channelDone)
+	case <-p.beatDone:
+	}
+	// Waits until channel go-routine properly stopped
 	p.channelWg.Wait()
-	p.wg.Wait()
 }
 
 // createHarvester creates a new harvester instance from the given state
 func (p *Prospector) createHarvester(state file.State) (*harvester.Harvester, error) {
 
+	outlet := channel.NewOutlet(p.beatDone, p.harvesterChan, p.eventCounter)
```

Comment:
ah, I see... one outlet passed to harvester. If NewHarvester would get an interface instead of channel.Outlet, the wrapping could be done in the harvester. That is, the logic `OnEvent` and `OnEventSignal` will become more local to the harvester (no need for public definition).